au4/src/effects/builtin/tonegen/toneeffect.h

```diff
@@ -4,15 +4,37 @@
 #pragma once
 
 #include "libraries/lib-builtin-effects/ToneGenBase.h"
+#include "../common/generatoreffect.h"
+
+struct EffectSettings;
 
 namespace au::effects {
-class ToneEffect : public ::ToneGenBase
+class ToneEffect : public ::ToneGenBase, public GeneratorEffect
 {
 public:
     ToneEffect();
 
+    enum class Waveform
+    {
+        sine,
+        square,
+        sawtooth,
+        squareNoAlias,
+        triangle,
+    };
+
     static const ComponentInterfaceSymbol Symbol;
 
     ComponentInterfaceSymbol GetSymbol() const override;
+
+    Waveform waveform() const;
+    void setWaveform(Waveform waveform);
+    double amplitude() const;
+    void setAmplitude(double amplitude);
+    double frequency() const;
+    void setFrequency(double frequency);
```

Comment:
we need to decide how important it is to return the boundary values ​​(min, max) here, i.e. who is responsible for them, the effect itself (it says I work within these limits) or this knowledge does not belong to anyone and is indicated where needed, for example in UI controls. 
(about amplitude and frequency)

Reply:
I don't understand this comment, to begin with, why you mention min and max.

Reply:
This effect can accept amplitude within certain limits, minimum - 0.0, maximum - 1.0 (not 2.0, not 20000.0 ...)
As I understand it, the frequency also has limits?
So we need to decide who is responsible for these limits.

For example, in Amplify these limits are set in the effect itself.
`AmplifyBase.h`
```
   static constexpr EffectParameter Ratio {
      &AmplifyBase::mRatio, L"Ratio", 0.9f, 0.003162f, 316.227766f, 1.0f
   };
```

```
Param<db_t> AmplifyEffect::amp() const
{
    return make_param<db_t>(muse::linear_to_db(mRatio), muse::linear_to_db(Ratio.min), muse::linear_to_db(Ratio.max));
}
```